fe/estimator.py

```diff
@@ -99,29 +135,41 @@ def _deltaG(model, sys_params):
 
         results = [x.result() for x in futures]
 
-    du_dls = []
+    if not (callback is None):
+        # TODO: what should the signature of callback be?
+        callback(results)
+
+    mean_du_dls = []
     all_grads = []
 
-    for (du_dl_mean, du_dl_std), grads in results:
-        # (ytz): figure out what to do with du_dl_std later
-        du_dls.append(du_dl_mean)
-        all_grads.append(grads)
+    for result in results:
+        # (ytz): figure out what to do with stddev(du_dl) later
+        mean_du_dls.append(np.mean(result.du_dls))
+        all_grads.append(result.du_dps)
 
-    dG = np.trapz(du_dls, model.lambda_schedule)
+    dG = np.trapz(mean_du_dls, model.lambda_schedule)
     dG_grad = []
     for rhs, lhs in zip(all_grads[-1], all_grads[0]):
         dG_grad.append(rhs - lhs)
 
     return dG, dG_grad
 
-@functools.partial(jax.custom_vjp, nondiff_argnums=(0,))
-def deltaG(model, sys_params):
-    return _deltaG(model, sys_params)[0]
+@functools.partial(jax.custom_vjp, nondiff_argnums=(0,2))
+def deltaG(model, sys_params, callback=None) -> float:
+    return _deltaG(model=model, sys_params=sys_params, callback=callback)[0]
 
-def deltaG_fwd(model, sys_params):
-    return _deltaG(model, sys_params)
+def deltaG_fwd(model, sys_params, callback=None) -> Tuple[float, gradient]:
```

Comment:
Someone posted a reply in this issue, so we might be able to make the aux stuff less janky
```
https://github.com/google/jax/issues/3806
```


``` python
import dataclasses
import jax
import jax.numpy as jnp

@dataclasses.dataclass
class Foo:
    jax_thing: jnp.array
    non_jax_thing: int

def flatten(foo):
    return (foo.jax_thing,), (foo.non_jax_thing,)

def unflatten(aux_data, children):
    non_jax_thing, = aux_data
    jax_thing, = children
    return Foo(jax_thing, non_jax_thing)

jax.tree_util.register_pytree_node(Foo, flatten, unflatten)
```